### ISSUE TITLE
Font does not support ANSI character range

### ISSUE THREAD

**Issue by tomek** [mintty/mintty#527]
![ss 2016-03-19 at 02 50 03](https://cloud.githubusercontent.com/assets/277443/13899446/aca883fe-edee-11e5-8261-b2ccf42b73b9.png)

Mintty 2.2.4 on cygwin64, Windows 10x64
Locale set to 'C', Character set 'UTF-8', I've 'Show bold as colour' + 'Allow blinking' enabled + Font smoothing at Full

This happens with numerous fonts like: Anonymous Pro (look @ img), Consolas, DejaVu Sans Mono, Lucida Console, Source Code Pro, Ubuntu Mono and this didn't happen in previous Mintty. I didn't notice anything wrong when I click 'ok' - it is just annoying to do this each time I start Mintty. 

I didn't check all my fonts but this doesn't happen with Iosevka font.

**Comment by mintty** [comment]
The hint about Iosevka is interesting. Anyway, what language version of Windows do you have?

**Comment by tomek** [comment]
Polish.

I checked if the same thing happens when I change locale to 'pl_Pl' or 'Defaut' - yep, same issue.

**Comment by elieux** [comment]
Happens to me on MSYS2 as well, starting with v2.2.4 (Windows 8.1x64). I have tried "Consolas" and "Lucida Console", happens with both.

I have my ANSI locale set to Czech, but changing it to English with [AppLocale](https://en.wikipedia.org/wiki/AppLocale) didn't help (but it's not 100 percent reliable, so it's not definitive evidence).

Is there anything I can do to debug this?

**Comment by fracting** [comment]
I just test Wine + mintty from latest git HEAD, I got an error message of "Font not found, using system substitute", which is understandable because Lucida Console is a Windows only font and not exist on Wine. I installed Lucida Console by `winetricks lucida`, and this time it complains "Font does not support system locale", this doesn't happen in mintty 2.2.4.

I'll test on Windows and report back.

**Comment by fracting** [comment]
Confirming same "Font does not support system locale" warning on Windows XP with Chinese locale.

**Comment by fracting** [comment]
Wine's TranslateCharsetInfo does not support TCI_SRCLOCALE, this is an unimplemented feature of Wine.

On Chinese version of Windows XP, get_default_charset() returns 134, however, lfp->lfCharSet is 1, which is DEFAULT_CHARSET defined in include/wingdi.h.

Any idea what's wrong?

Thanks!

**Comment by mintty** [comment]
It seems to be a limitation of Wine, too, if lfp->lfCharSet.
I have added some precautions for the two cases you describe, and with a Chinese setting, the warning is now shown only for those fonts that actually miss support for the selected system locale. But maybe the warning isn't really desirable in that case either? Opinions, please, so I'll introduce an option or disable the warning alltogether.

**Comment by fracting** [comment]
The current solution works for me, it builds and pass the unattended test now: https://tea-ci.org/fracting/mintty/21

I'm happy with the current solution. For me, what I care is whether there is a way to run everything in command line without human clicking or so. So, either a command parameter to silent the dialog or totally disable the warning will work the same good for me. However, since the latest patch fix the warning, so it is already good enough for me, thanks a lot for the quick response!

**Comment by elieux** [comment]
Not sure if I'm supposed to create a new ticket for this (v2.3.1):
```
windres --preprocessor 'gcc -E -xc -DRC_INVOKED -MMD -MP -DTARGET=x86_64-pc-msys -DNDEBUG' res.rc ../bin/msys64/res.o
wintext.c: In function ‘get_default_charset’:
wintext.c:205:28: error: cast to pointer from integer of different size [-Werror=int-to-pointer-cast]
   if (TranslateCharsetInfo((DWORD *)GetSystemDefaultLCID(), &csi, TCI_SRCLOCALE))
```

**Comment by mintty** [comment]
Fixed in 2.3.2.
Except people might not appreciate the warning if valid; in that case we'll see...

**Comment by antarestrue** [comment]
I am still hitting this warning on v2.3.2: "Font does not support system locale"
The font is Consolas;
the locale ($LANG) is set to zh_TW.utf8@cjknarrow (en_US.utf8 doesn't suppress the warning, too);
Windows is Win7 English version but locale set to Taiwan.

I need to unset font to suppress the warning, but the default font is so ugly.

Is there any way to get rid of this warning?  My current setting has been working well all the time.  I don't see a reason for that warning message.

**Comment by antarestrue** [comment]
default_charset here is 136.  DEFAULT_CHARSET is 1.
lfp->lfCharSet I have are 0, 161, 162, 186, 238, 204, 163.

**Comment by tomek** [comment]
@mintty thanks for your fix, now everything looks fine for me- I'm leaving closing this issue up to you - other people have other issues and I'm not sure if this one can be closed or not.

**Comment by antarestrue** [comment]
@mintty Thanks for the fix.  I don't see the warning now.
Just curious, doesn't Windows has the ability to choose a fallback font when a character is not found in the current font?  Then what's the point of this warning?

**Comment by mintty** [comment]
It’s quite mysterious how the font fallback would work; for a demo of missing characters try `mintty -o Font=Rod` or `mintty -o Font=Terminal` and output some ANSI characters (Dænish, Frènch, Spañish, Görman, Turkıș), or `mintty -o Font="WP MultinationalA Courier"`.

**Comment by hexiyou** [comment]
![mintty](https://cloud.githubusercontent.com/assets/1593944/13948286/34aa8b54-f05a-11e5-9df4-619cc52ab3ca.png)

**Comment by mintty** [comment]
The latest version of mintty has the warning changed to "`Font has limited support for character ranges`" and it is shown only if the font supports neither of the "ANSI" character range nor the current system locale. See my previous comment for some font examples.

**Comment by mintty** [comment]
Recent tweak (warning only on stderr if possible) released in 2.3.6.
